File: apps/api/app/schemas/imports.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.enums import (
    MarketDataFreshnessStatus,
    MarketDataSource,
    MarketDataStatus,
    MarketDataSyncStatus,
    Timeframe,
)


class CsvImportError(BaseModel):
    row: int | None = None
    field: str | None = None
    message: str


class CsvImportResult(BaseModel):
    series_id: int | None
    watchlist_item_id: int
    timeframe: Timeframe
    status: MarketDataStatus
    candle_count: int
    start_time: datetime | None
    end_time: datetime | None
    source: MarketDataSource
    freshness_status: MarketDataFreshnessStatus
    sync_status: MarketDataSyncStatus
    last_synced_at: datetime | None
    errors: list[CsvImportError]


class ImportHistoryItem(BaseModel):
    series_id: int
    watchlist_item_id: int
    symbol: str
    timeframe: Timeframe
    status: MarketDataStatus
    candle_count: int
    start_time: datetime | None
    end_time: datetime | None
    imported_at: datetime
    source: MarketDataSource
    freshness_status: MarketDataFreshnessStatus
    sync_status: MarketDataSyncStatus
    last_synced_at: datetime | None
    provider_name: str | None
    file_name: str | None


class MarketDataSyncRequest(BaseModel):
    watchlist_item_id: int
    timeframe: Timeframe


class ProviderTimeframeCapabilityRead(BaseModel):
    timeframe: Timeframe
    supported: bool
    reason: str


def default_alpha_vantage_capabilities() -> list[ProviderTimeframeCapabilityRead]:
    return [
        ProviderTimeframeCapabilityRead(
            timeframe=Timeframe.ONE_WEEK,
            supported=False,
            reason=(
                "Alpha Vantage first path does not support weekly sync here; "
                "use CSV fallback."
            ),
        ),
        ProviderTimeframeCapabilityRead(
            timeframe=Timeframe.ONE_DAY,
            supported=True,
            reason="Guarded Daily/EOD sync path is supported for configured symbols.",
        ),
        ProviderTimeframeCapabilityRead(
            timeframe=Timeframe.FOUR_HOURS,
            supported=False,
            reason=(
                "4H/intraday provider sync is not selected; "
                "use TradingView CSV fallback."
            ),
        ),
    ]


class MarketDataSyncResponse(BaseModel):
    watchlist_item_id: int
    series_id: int
    source: MarketDataSource
    timeframe: Timeframe
    candle_count: int
    provider_name: str | None
    provider_symbol: str | None
    provider_exchange: str | None
    provider_timeframe: str | None
    sync_status: MarketDataSyncStatus
    freshness_status: MarketDataFreshnessStatus
    last_synced_at: datetime | None
    end_time: datetime | None
    sync_error_code: str | None
    sync_error_message: str | None
    capability_note: str = (
        "Provider sync is stored-data only. Unsupported timeframes require CSV fallback."
    )
    provider_capabilities: list[ProviderTimeframeCapabilityRead] = Field(
        default_factory=default_alpha_vantage_capabilities
    )

File: apps/web/src/types/imports.ts
export type Timeframe = "1W" | "1D" | "4H";

export type MarketDataStatus = "imported" | "validated" | "failed" | "analyzed";

export type MarketDataSource = "tradingview_csv" | "manual" | "api_later" | "provider" | "unknown";

export type MarketDataFreshnessStatus = "fresh" | "stale" | "unknown" | "failed" | "partial";

export type MarketDataSyncStatus = "not_applicable" | "success" | "skipped" | "failed" | "partial";

export type SignalStatus = "watchlist" | "armed" | "triggered" | "invalidated" | "no_setup" | "missed" | "expired";

export type ScoreClass = "a_setup" | "b_setup" | "watchlist" | "no_trade";

export type Bias = "bullish" | "neutral" | "bearish";

export type StrategyType = "trend_pullback_long" | "base_breakout_long";

export type CsvImportError = {
  row: number | null;
  field: string | null;
  message: string;
};

export type CsvImportResult = {
  series_id: number | null;
  watchlist_item_id: number;
  timeframe: Timeframe;
  status: MarketDataStatus;
  candle_count: number;
  start_time: string | null;
  end_time: string | null;
  source: MarketDataSource;
  freshness_status: MarketDataFreshnessStatus;
  sync_status: MarketDataSyncStatus;
  last_synced_at: string | null;
  errors: CsvImportError[];
};

export type ImportHistoryItem = {
  series_id: number;
  watchlist_item_id: number;
  symbol: string;
  timeframe: Timeframe;
  status: MarketDataStatus;
  candle_count: number;
  start_time: string | null;
  end_time: string | null;
  imported_at: string;
  source: MarketDataSource;
  freshness_status: MarketDataFreshnessStatus;
  sync_status: MarketDataSyncStatus;
  last_synced_at: string | null;
  provider_name: string | null;
  file_name: string | null;
};

export type MarketDataSyncRequest = {
  watchlist_item_id: number;
  timeframe: Timeframe;
};

export type MarketDataSyncResult = {
  watchlist_item_id: number;
  series_id: number;
  source: MarketDataSource;
  timeframe: Timeframe;
  provider_name: string | null;
  provider_symbol: string | null;
  provider_exchange: string | null;
  provider_timeframe: string | null;
  sync_status: MarketDataSyncStatus;
  freshness_status: MarketDataFreshnessStatus;
  last_synced_at: string | null;
  end_time: string | null;
  sync_error_code: string | null;
  sync_error_message: string | null;
};

export type SignalAnalysisResult = {
  strategy_type: StrategyType;
  status: SignalStatus;
  bias: Bias;
  score: number;
  score_class: ScoreClass;
  timeframe_context: Timeframe;
  timeframe_setup: Timeframe;
  timeframe_trigger: Timeframe;
  entry_low: string | null;
  entry_high: string | null;
  trigger_level: string | null;
  stop_loss: string | null;
  target_1: string | null;
  target_2: string | null;
  risk_reward: string | null;
  invalidation_reason: string | null;
  reasoning: string[];
  risk_flags: string[];
  next_action: string;
  no_trade_reasons: string[];
  quality_report: AnalysisQualityCheck[];
};

export type AnalysisQualityCheck = {
  key: string;
  label: string;
  status: "passed" | "warning" | "blocked" | "missing" | string;
  detail: string;
};

export type MarketDataAnalysisResult = {
  series_id: number;
  status: MarketDataStatus;
  candle_count: number;
  indicator_snapshot_count: number;
  signal: SignalAnalysisResult;
};
